src/hd/usb.c

```diff
@@ -569,41 +569,48 @@ void add_input_dev(hd_data_t *hd_data, char *name)
       s
     );
 
-    for(hd = hd_data->hd; hd; hd = hd->next) {
-      if(
-        hd->module == hd_data->module &&
-        hd->sysfs_id &&
-        s &&
-        !strcmp(s, hd->sysfs_id)
-      ) {
-        t = NULL;
-        str_printf(&t, 0, "/dev/input/%s", sf_cdev_name);
+    /* find device (matching sysfs path) */
+    hd = hd_find_sysfs_id(hd_data, s);
+
+    /* not found? - retry one level up */
+    if(!hd) {
+      char *ns = new_str(s), *nt;
+      if((nt = strrchr(ns, '/'))) {
+        *nt = 0;
```

Comment:
ok, now I get code. quite obfuscated.
what about something like
```
nt = strrchr(s, '/');
unsigned substr_len = nt - s;
char ns = malloc(substr_len + 1);
memcpy(ns, s, substr_len);
ns[substr_len] = 0;
```

this way it is a bit longer, but for me more readable. as it is obvious that it copy substring to new memory location. Also can mem_free dealocate whole string even when earlier finished? ( I do not check implementation if it rely on string or if it use memory allocation map )